core/src/ctx/context.rs

```diff
@@ -71,6 +73,9 @@ pub struct MutableContext {
 	transaction: Option<Arc<Transaction>>,
 	// Does not read from parent `values`.
 	isolated: bool,
+	// Idiom recursion details
+	#[allow(clippy::type_complexity)]
+	idiom_recursion: Option<(Arc<Mutex<i64>>, i64, Recurse, Vec<Part>)>,
```

Comment:
This might be personal preference but I dislike storing parts of the AST inside the context. Looks to my like breaking a separation of concerns and it introduces extra mutable state. 

Could we maybe solve the issue of issue of having to restart evaluating an idiom with a error type? Similar to how return and break statements do it? That way we can also don't use extra stack.

Reply:
Sure, I can separate it out. I figured the type was only used in a single place, but a separate type will be cleaner indeed.

Let's first take an example:
```sql
person:tobie.{3..}.{ id, knows: ->knows->person.@ };
```

Here, `{3..}` starts the recurse and stores stores `.{ id, knows: ->knows->person.@ }` inside the context. When we later process the `RepeatRecurse` (`.@`) symbol, we need the context of that path, to repeat the pattern.

The Arc<Mutex> is there because nested recursions need to be able to store how deep we iterated. A path may go two iterations deep, in which case the result should be omitted as we requested at least 3 iterations.

I tried to go down the error route, however then we may not walk all possible paths.
Let's say that `->knows->person` results in `[person:jaime, person:lizzie]`, then the RepeatRecurse symbol will first process the path of `person:jaime`. At some point we would get back an error like `Error::Recursed` with some details of how deep we went, what the result is, is there more path we need to process, where in what value do we store the result. 
First of all, this will add a lot of logic to structure this back to a proper output, secondly we can no longer process `person:lizzie` as we reached an error.

I tried to go down this path first, but it became way too complex way to fast, it's not feasible. I do agree this adds a cost when you are actually recursing, but relying on `.get()` with a bit of context for actual recursion simply seems like the better solution here.

Keep in mind that this only applies to paths which actually need to recurse, like when you use the RepeatRecurse symbol inside a destructure. The following examples will not recurse as an example and will instead rely on the loop to collect a result:
```sql
person:tobie.{3..}->knows->person
person:tobie.{3..}->knows->person.@
person:tobie.({3..}->knows->person).name
person:tobie.{3..}->knows->person.@.name
```

These also give very different outputs. The first example is used to create a arbitrary depth recursive tree, while the second code snippet will result in a single array which collects all outputs.

Reply:
I'm not actually sure if this Arc<Mutex> counter currently results in the correct behaviour. For a query like:
```sql
person:tobie.{3..}.{ id, knows: ->knows->person.@ };
```

This would be a valid output:
```sql
{
  id: person:tobie,
  knows: [
    {
      id: person:jaime,
      knows: NONE
    },
    {
      id: person:lizzie,
      knows: [
        {
          id: person:micha,
          knows: NONE
        }
      ]
    }
  ]
}
```

Because one of the paths recursed to the required depth, but the paths that do not reach that level of recursion are not currently eliminated, I think.

Reply:
Agree with keeping the separation of concerns.
Also, normally, context's mutability is achieved by creating new child contexts.

Query Planner, Full-Text, HNSW Indexes and KNN brute forces are also examples of mutable and non-mutable states maintained during the query execution.
In this case, the `Context` provides an access to the state. (get_iteration_stage, get_query_planner, etc...)

Also, wouldn't an AtomicInteger do the job?

Reply:
I think I may have a solution, I'm not yet sure if I can elimitate the Arc<Mutex>, but potentially. We definitely do need the context however, simply because the RepeatRecurse symbol otherwise does not have any context

Reply:
Fixed in [e2495d7](https://github.com/surrealdb/surrealdb/pull/5086/commits/e2495d71685f63b6f4c38abf799e760265b355d4)